### PR TITLE
lite: enable (u)int8 quantization and int32 for ABS

### PR DESCRIPTION
**Pull request by zhenhuaw-me**
Hi, this PR adds uint8/int8/int32 support for ABS kernel in TFLite. Detail change includes:
1. Move ABS out routine and kernel from elementwise generic as ops have different quantization handling.
2. Float/Int32 ABS invoke `std::abs` directly.
3. Int8/UInt8 ABS involving requantization after `std::abs`.
4. Float as version 1, Int32/Int8/UInt8 as version 2.
5. Test for kernel and TOCO.

### CODE REVIEW

**Comment by jackwish** [comment]
One open discussion is that, do we really need requantization in quantized ABS.

Intuitively, it's not needed since the representation of FP32 in INT8 won't change, that is the OP's property. However, we can safely remove the requantize only when TOCO enforcing the same quantization representation for input and output tensor, which seems is not the case currently. We have a model which has different quantization representation of ABS input/output when converted to TFLite model. So maybe, we can investigate that should TOCO (and any similar quantization tools) enforce such check when converting models.

Would you please help to CC people who may have interest or expertise in such topic (I mean where need to evaluate quantization representation handling in model conversion)? Looking forward to hear from you guys. Thank you.

**Comment by jackwish** [comment]
Hi there, any update needed?

**Comment by jackwish** [comment]
Two more commits to extend Quantization-aware Training to operator Sub, Abs, Squeeze, ExpandDims and so on; and remove requantize in Abs TFLite kernel as it should have same quantization representation.

Also loop @aselle and @suharshs who may have interest.

File: tensorflow/lite/kernels/abs.cc [hunk 1]
```diff
@@ -0,0 +1,136 @@
+/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.
```
Comment (suharshs) [inline tensorflow/lite/kernels/abs.cc]: nit: could you make these 2019

File: tensorflow/lite/kernels/internal/types.h [hunk 2]
```diff
@@ -758,6 +758,12 @@ struct ArithmeticParams {
   int broadcast_shape[5];
 };
 
+struct AbsParams {
+  // uint8 inference params.
```
Comment (suharshs) [inline tensorflow/lite/kernels/internal/types.h]: uint8/int8

File: tensorflow/lite/kernels/test_util.h [hunk 3]
```diff
@@ -65,6 +65,12 @@ inline std::vector<float> Dequantize(const std::vector<T>& data, float scale,
   return f;
 }
 
+float GetQuantizeTolerance(int min, int max) {
+  float QuantizedStep = (max - min) / 255.0;
```
Comment (suharshs) [inline tensorflow/lite/kernels/test_util.h]: nit: Name QuantizedStep quantized_step and same for other names.

File: tensorflow/lite/toco/graph_transformations/hardcode_min_max.cc [hunk 4]
```diff
@@ -430,6 +430,7 @@ ::tensorflow::Status HardcodeMinMax::Run(Model* model, std::size_t op_index,
       changed = HardcodeMinMaxForAverageOrMaxPool(model, op);
       break;
 
+    case OperatorType::kAbs:
```
Comment (suharshs) [inline tensorflow/lite/toco/graph_transformations/hardcode_min_max.cc]: is this hardcode case intended? This hardcodes the min/max of the output to the min/max of the input. this means if the abs input is [-2, -1] than the output range will be the same, which seems off. 

Couple options:
(1) Choose to not have a hardcode case here, instead relying on a Fakequant from contrib/quatnize?
(2) Hardcode to:
r = max(abs(input_min), (input_max))
[-r, r]

WYDT?

File: tensorflow/lite/toco/tflite/operator_test.cc [hunk 5]
```diff
@@ -158,6 +158,37 @@ TEST_F(OperatorTest, SimpleOperators) {
...
+TEST_F(OperatorTest, BuiltinAbs) {
+  AbsOperator abs_op;
+  abs_op.inputs = {"input"};
+  auto operator_by_type_map = BuildOperatorByTypeMap(false /*enable_flex_ops*/);
+  const BaseOperator* op = operator_by_type_map.at(abs_op.type).get();
+
+  Model float_model;
+  Array& input_float_array = float_model.GetOrCreateArray(abs_op.inputs[0]);
+  input_float_array.data_type = ArrayDataType::kFloat;
+  OperatorSignature float_signature = {.op = &abs_op, .model = &float_model};
```
Comment (suharshs) [inline tensorflow/lite/toco/tflite/operator_test.cc]: unfortunately this .op notation breaks on come compilers :(

File: tensorflow/lite/toco/tflite/operator.cc [hunk 6]
```diff
@@ -182,6 +182,22 @@ class DepthwiseConvolution
...
+class Abs : public SimpleOperator<FloorDivOperator> {
+ public:
+  explicit Abs() : SimpleOperator("ABS", OperatorType::kAbs) {}
+  int GetVersion(const OperatorSignature& op_signature) const override {
+    const string& input_name = op_signature.op->inputs[0];
+    const Array& input_array = op_signature.model->GetArray(input_name);
+    // Version 2 supports signed/unsigned int8 and signed int32 input types.
+    if (input_array.data_type == ArrayDataType::kInt8 ||
+        input_array.data_type == ArrayDataType::kUint8 ||
+        input_array.data_type == ArrayDataType::kInt32) {
```
Comment (suharshs) [inline tensorflow/lite/toco/tflite/operator.cc]: thanks! please update version in op resolver too https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/register.cc#L161

File: tensorflow/lite/toco/tflite/operator.cc [hunk 7]
```diff
@@ -1060,6 +1076,8 @@ class Lstm : public BuiltinOperator<LstmCellOperator, ::tflite::LSTMOptions,
       case LstmCellOperator::KERNEL_BASIC:
         // KERNEL_BASIC was added in version 2.
         return 2;
+      default:
+        return -1;
```
Comment (suharshs) [inline tensorflow/lite/toco/tflite/operator.cc]: this was intentionally left out to have a compiler error if we miss an case

File: tensorflow/contrib/quantize/python/quantize.py [hunk 8]
```diff
@@ -37,10 +37,10 @@
 _RELU_TYPES = {'Relu', 'Relu6'}
 
 _QUANTIZATION_OP = {'FakeQuantWithMinMaxVars'}
-_VALID_SRC_OP = {'Add', 'AddV2', 'Mul'}
-_INTERMEDIATE_OP = {'Add', 'AddV2', 'Mul'}
+_VALID_SRC_OP = {'Add', 'AddV2', 'Mul', 'Sub', 'ConcatV2'}
```
Comment (suharshs) [inline tensorflow/contrib/quantize/python/quantize.py]: could we add a test for the contrib/quantize changes, and move the contrib/quantize changes to a separate pr from the tflite changes.

**Review by suharshs — COMMENTED** [review]
Awesome work!

**Comment by suharshs** [comment]
I will spend some time to think more about the requantization case for abs, for now i think going to int32 like you are doing is fine. Thanks!

File: tensorflow/lite/kernels/abs.cc (hunk 1 above)
Comment (jackwish) [reply tensorflow/lite/kernels/abs.cc]: Sure, thank you!

File: tensorflow/lite/kernels/internal/types.h (hunk 2 above)
Comment (jackwish) [reply tensorflow/lite/kernels/internal/types.h]: Yes, will update.

File: tensorflow/lite/kernels/test_util.h (hunk 3 above)
Comment (jackwish) [reply tensorflow/lite/kernels/test_util.h]: Yes, will update!

File: tensorflow/lite/toco/graph_transformations/hardcode_min_max.cc (hunk 4 above)
Comment (jackwish) [reply tensorflow/lite/toco/graph_transformations/hardcode_min_max.cc]: Yes, it **was** intended. Like mentioned https://github.com/tensorflow/tensorflow/pull/31914#issuecomment-524169853, I have been choosing whether to let input/output of Abs having different quantization representation. But I really missed the value range that you pointed out. This is not what we want.

File: tensorflow/lite/toco/tflite/operator_test.cc (hunk 5 above)
Comment (jackwish) [reply tensorflow/lite/toco/tflite/operator_test.cc]: Ok then.. I was copying from some other tests...

File: tensorflow/lite/toco/tflite/operator.cc (hunk 6 above)
Comment (jackwish) [reply tensorflow/lite/toco/tflite/operator.cc]: Sure. I am thinking that do we need a guide to let people know the *must* of adding operators related code to TFLite? I believe many people have similar questions like  https://github.com/tensorflow/tensorflow/issues/31879

File: tensorflow/lite/toco/tflite/operator.cc (hunk 7 above)
Comment (jackwish) [reply tensorflow/lite/toco/tflite/operator.cc]: Got it, I will leave some comments here :)

File: tensorflow/contrib/quantize/python/quantize.py (hunk 8 above)
Comment (jackwish) [reply tensorflow/contrib/quantize/python/quantize.py]: Sure, I will add some tests for this.

Regarding moving the `contrib/quantize` out, I am not sure about setting min/max of ABS in TOCO, which should be consistent with the TFLite kernel I think.

**Comment by jackwish** [comment]
> I will spend some time to think more about the requantization case for abs, for now i think going to int32 like you are doing is fine. Thanks!

Thank you for the review @suharshs !  For the Abs requantization, I was paying too much attention on the representation precision about input/output - for most case overlapping zero, the output range is smaller than the input range, but the data representation cannot be more accurate in output. I mean, the information of output cannot be more than input, so I thought a requantization is not needed. Leading to miss the range case `[-2, -1]` you pointed out, for which a requantization seems needed. Thank you for the patient review!

The min/max design choice also impacts `contrib/quantize`, where I took ABS as *pass through* operators - while should not. I think learn min/max for Abs rather than hardcode it is a good idea, just like what was done for Concat in this PR.

Also, out of the scope for this PR, I am thinking that the open sourced Quantization-aware Training may need redesign. Current design basically takes operators which holding weights as something like *master* node, and processes based on them. Many more operators, such as the already added `Add` and `Mul`, can be improved by QAT. I guess we can add FakeQuant to anywhere, unless it is not necessary - `Reshape` for example. We may let TensorFlow operators holding some attributions describing where the quantization representation of their inputs/outputs may be different, and the `contrib/quantize` can simply walk the network and insert FakeQuant nodes. That is what I am thinking, as someone who don't hold a full sight of the design. I will be very happy to contribute such things. Hoping it were not going too far wrong. :)

I will update the PR next week to let Abs learn min/max. I am thinking that the `contrib/quantize` and tflite part code can be in one PR to let them have consistent behavior on quantization related things (int32 is really insignificant). What do you think?

File: tensorflow/lite/toco/graph_transformations/hardcode_min_max.cc (hunk 4 above)
Comment (jackwish) [reply tensorflow/lite/toco/graph_transformations/hardcode_min_max.cc]: Removed Abs from hardcode, maybe we can learn min/max in further patch.

File: tensorflow/contrib/quantize/python/quantize.py (hunk 8 above)
Comment (jackwish) [reply tensorflow/contrib/quantize/python/quantize.py]: Removed `contrib/quantize` changes.

**Comment by jackwish** [comment]
Removed `contrib/quantize` changes from this PR. Will open another for it. Please review again @suharshs .

**Review by suharshs — DISMISSED** [review]

**Comment by gbaned** [comment]
@jackwish Could you please check build failures and resolve the conflicts? Thanks!

**Comment by jackwish** [comment]
> @jackwish Could you please check build failures and resolve the conflicts? Thanks!

Hi @gbaned , conflicts has been fixed, thank you for the reminder.
Hi @suharshs , could you please have another look in this?

**Comment by gbaned** [comment]
@jackwish Could you please resolve the conflicts? Thanks!

**Comment by jackwish** [comment]
Hi @gbaned the conflicts have been fixed. The latest two rounds of conflict resolving are due to the long progress of this PR. Hoping that we can accelerate the review progress before new conflicts emerge :)

**Comment by gbaned** [comment]
@jackwish Sorry for the slow response. Could you please resolve the conflicts? Thanks!

**Review by suharshs — DISMISSED** [review]

**Comment by gbaned** [comment]
@jackwish Could you please check build failures and resolve the conflicts? Thanks!

**Comment by jackwish** [comment]
Hi @gbaned the conflicts have been fixed, and @suharshs @wangtz @jianlijianli please have another look.

PS. As tree round of conflicts resolving has been conducted in this PR, I am now unsubscribing from this PR thread and leaving it as it is. Thank you guys for all the effort, best regards.

**Comment by gbaned** [comment]
@jackwish Could you please resolve the conflicts? Thanks!

**Comment by jackwish** [comment]
> @jackwish Could you please resolve the conflicts? Thanks!

Thank you @gbaned for the remind.

But the conflicts resolving will not happen. As https://github.com/tensorflow/tensorflow/pull/31914#issuecomment-544493621 has stated, this PR is in a *conflicts resolving* loop which I will no longer pay efforts. Please feel free to close this PR if no further actions are going to happen from your side.

Thank you everyone for your time.

**Comment by jackwish** [comment]
Now closing this PR. Thanks everyone for your time.

I will save the time of contributing back to tensorflow official repo to maintain a quantization enhanced fork based on TensorFlow 1.14. For anyone interested in this PR or any further related features, please switch to [my repo](https://github.com/jackwish/tensorflow/tree/lts/r1.14/).

The enhancement covers two aspects basically:
* More quantized operators and kernels in TensorFlow Lite.
* More scenarios for Quantization-aware Training.

**Comment by jackwish** [comment]
switch to #34263

**Pull request closed without merging**